File: take_u_forward/graph/15_Directed Graph Cycle.java
import java.util.*;

class Solution {

    private boolean dfsCheck(int node,
                             ArrayList<ArrayList<Integer>> adj,
                             int[] vis,
                             int[] pathVis) {

        vis[node] = 1;
        pathVis[node] = 1;

        // Traverse adjacent nodes
        for (int it : adj.get(node)) {

            // If node is not visited
            if (vis[it] == 0) {
                if (dfsCheck(it, adj, vis, pathVis)) {
                    return true;
                }
            }
            // If node is visited and in current path
            else if (pathVis[it] == 1) {
                return true;
            }
        }

        pathVis[node] = 0; // backtrack
        return false;
    }

    public boolean isCyclic(int V, int[][] edges) {

        // Create adjacency list
        ArrayList<ArrayList<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < V; i++) {
            adj.add(new ArrayList<>());
        }

        // Build graph
        for (int[] e : edges) {
            adj.get(e[0]).add(e[1]);
        }

        int[] vis = new int[V];
        int[] pathVis = new int[V];

        // Run DFS for each component
        for (int i = 0; i < V; i++) {
            if (vis[i] == 0) {
                if (dfsCheck(i, adj, vis, pathVis)) {
                    return true;
                }
            }
        }

        return false;
    }
}

File: take_u_forward/graph/15_Directed Graph Cycle.cpp
class Solution {
private:
    bool dfsCheck(int node, vector<vector<int>> &adj,
                  vector<int> &vis, vector<int> &pathVis) {

        vis[node] = 1;
        pathVis[node] = 1;

        for (int it : adj[node]) {
            // If node is not visited
            if (!vis[it]) {
                if (dfsCheck(it, adj, vis, pathVis))
                    return true;
            }
            // If node is visited and present in current DFS path
            else if (pathVis[it]) {
                return true;
            }
        }

        pathVis[node] = 0;  // backtrack
        return false;
    }

public:
    bool isCyclic(int V, vector<vector<int>> &edges) {
        // Create adjacency list
        vector<vector<int>> adj(V);
        for (auto &e : edges) {
            adj[e[0]].push_back(e[1]);
        }

        vector<int> vis(V, 0);
        vector<int> pathVis(V, 0);

        for (int i = 0; i < V; i++) {
            if (!vis[i]) {
                if (dfsCheck(i, adj, vis, pathVis))
                    return true;
            }
        }
        return false;
    }
};
